src/executables/csql.c

```diff
@@ -3330,3 +3358,168 @@ get_host_ip (unsigned char *ip_addr)
+
+#if defined(CS_MODE)
+  memset (boot_Host_connected, 0, sizeof (boot_Host_connected));
+#endif /* CS_MODE */
+
+  /*Failed to access other host or db and then access formal db_name */
+  if (db_name_ptr == NULL)
+    {
+      db_name_ptr = strdup (csql_arg->db_name);
+    }
+
+  if (csql_Database_connected)
+    {
+      csql_Database_connected = false;
+      db_end_session ();
+      db_shutdown ();
+
+      fprintf (csql_Output_fp, "Warning: current CSQL session is disconnected.\n");
+    }
+
+  er_init ("./csql.err", ER_NEVER_EXIT);
+  if (db_restart_ex (UTIL_CSQL_NAME, db_name_ptr, user_name_ptr, NULL, NULL, db_get_client_type ()) != NO_ERROR)
+    {
+      if (csql_Is_interactive && db_error_code () == ER_AU_INVALID_PASSWORD)
+	{
+	  p = getpass ((char *) csql_get_message (CSQL_PASSWD_PROMPT_TEXT));
+
+	  /* try again */
+	  if (db_restart_ex (UTIL_CSQL_NAME, db_name_ptr, user_name_ptr, p, NULL, db_get_client_type ()) != NO_ERROR)
+	    {
+	      csql_Error_code = CSQL_ERR_SQL_ERROR;
+	      csql_check_server_down ();
+
+	      return DO_CMD_FAILURE;
+	    }
+	  else
+	    {
+	      /*login success with password */
+	      csql_new_arg.user_name = strdup (user_name_ptr);
+	      csql_new_arg.db_name = strdup (db_name_ptr);
+
+	      if (p[0] == '\0')
+		{
+		  csql_new_arg.passwd = (char *) NULL;	/* to fit into db_login protocol */
+		}
+	      else
+		{
+		  csql_new_arg.passwd = strdup (p);
+		}
+
+	      FREE_MEM ((char *) csql_arg->user_name);
+	      FREE_MEM ((char *) csql_arg->db_name);
+	      FREE_MEM ((char *) csql_arg->passwd);
+
+	      memcpy (csql_arg, &csql_new_arg, sizeof (CSQL_ARGUMENT));
+
+	      if (csql_arg->sysadm && au_is_dba_group_member (Au_user))
+		{
+		  au_sysadm_disable ();
+		}
+	      csql_Database_connected = true;
+
+	      if (csql_arg->trigger_action_flag == false)
+		{
+		  db_disable_trigger ();
+		}
+
+	      fprintf (csql_Output_fp, "CONNECTED : %s %s\n", csql_arg->user_name, csql_arg->db_name);
+//            fprintf (csql_Output_fp, "Connected.\n");
+	    }
+	}
+      else
+	{
+	  csql_Error_code = CSQL_ERR_SQL_ERROR;
+
+	  return DO_CMD_FAILURE;
+
+	}
+    }
+  else
+    {
+      /*login success without password */
+      csql_new_arg.user_name = strdup (user_name_ptr);
+      csql_new_arg.db_name = strdup (db_name_ptr);
+      csql_new_arg.passwd = (char *) NULL;
+
+      FREE_MEM ((char *) csql_arg->user_name);
+      FREE_MEM ((char *) csql_arg->db_name);
+      FREE_MEM ((char *) csql_arg->passwd);
+
+      memcpy (csql_arg, &csql_new_arg, sizeof (CSQL_ARGUMENT));
+
+      if (csql_arg->sysadm && au_is_dba_group_member (Au_user))
+	{
+	  au_sysadm_disable ();
+	}
+      csql_Database_connected = true;
+
+      if (csql_arg->trigger_action_flag == false)
+	{
+	  db_disable_trigger ();
+	}
+
+//            fprintf (csql_Output_fp, "Connected.\n");
+      fprintf (csql_Output_fp, "CONNECTED : %s %s\n", csql_arg->user_name, csql_arg->db_name);
+    }
+
```

Comment:
Line 3438 - 3523 contains same codes except csql_new_arg.passwd is NULL or not.
So, how about following code (this is just example):

```
er_init ("./csql.err", ER_NEVER_EXIT);
csql_new_arg.passwd = (char *) NULL;

 if (db_restart_ex (UTIL_CSQL_NAME, db_name_ptr, user_name_ptr, NULL, NULL, db_get_client_type ()) != NO_ERROR)
    {
      if (csql_Is_interactive && db_error_code () == ER_AU_INVALID_PASSWORD)
        {
          p = getpass ((char *) csql_get_message (CSQL_PASSWD_PROMPT_TEXT));

          /* try again */
          if (db_restart_ex (UTIL_CSQL_NAME, db_name_ptr, user_name_ptr, p, NULL, db_get_client_type ()) != NO_ERROR)
            {
              csql_Error_code = CSQL_ERR_SQL_ERROR;
              csql_check_server_down ();

              return DO_CMD_FAILURE;
            }
          else
            {
              /*login success with password */
              csql_new_arg.user_name = strdup (user_name_ptr);
              csql_new_arg.db_name = strdup (db_name_ptr);
            }

          csql_new_arg.passwd = (p[0] == NULL) ? (char *) NULL : strdup (p);
        }
    }

 csql_new_arg.user_name = strdup (user_name_ptr);
 csql_new_arg.db_name = strdup (db_name_ptr);

 FREE_MEM ((char *) csql_arg->user_name);
 FREE_MEM ((char *) csql_arg->db_name);
 FREE_MEM ((char *) csql_arg->passwd);

 memcpy (csql_arg, &csql_new_arg, sizeof (CSQL_ARGUMENT));
 if (csql_arg->sysadm && au_is_dba_group_member (Au_user))
   {
     au_sysadm_disable ();
   }
 csql_Database_connected = true;

 if (csql_arg->sysadm && au_is_dba_group_member (Au_user))
   {
     au_sysadm_disable ();
   }
 csql_Database_connected = true;

 if (csql_arg->trigger_action_flag == false)
   {
     db_disable_trigger ();
   }

//            fprintf (csql_Output_fp, "Connected.\n");
  fprintf (csql_Output_fp, "CONNECTED : %s %s\n", csql_arg->user_name, csql_arg->db_name);

return DO_CMD_SUCCESS;
```

Reply:
Thank you. I will apply it.

Reply:
After executing the ;connect command, only user_name, db_name, and passwd are changed. The code below is the same as before. So I think it's unnecessary code.

```
 if (csql_arg->sysadm && au_is_dba_group_member (Au_user))
   {
     au_sysadm_disable ();
   }

 if (csql_arg->trigger_action_flag == false)
   {
     db_disable_trigger ();
   }
```